### PR TITLE
New Expression: Loaded Chunks of World

### PR DESCRIPTION
**Pull request by DelayedGaming**
### Description
This PR adds support of getting [a list of loaded chunks of a world](https://jd.papermc.io/paper/1.19/org/bukkit/World.html#getLoadedChunks()) into Skript!

---
**Target Minecraft Versions:** any
**Requirements:** -
**Related Issues:** -

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprLoadedChunks.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.util.Kleenean;
+import org.bukkit.Chunk;
+import org.bukkit.World;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Loaded Chunks of World")
+@Description("A list of loaded chunks of a world.")
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprLoadedChunks.java:37]:
```suggestion
@Description("Returns a list of loaded chunks of a world.")
```

File: src/main/java/ch/njol/skript/expressions/ExprLoadedChunks.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+@Examples("set {_chunks::*} to the loaded chunks of player's world")
+@Since("INSERT VERSION")
+public class ExprLoadedChunks extends SimpleExpression<Chunk> {
+
+	static {
+		Skript.registerExpression(ExprLoadedChunks.class, Chunk.class, ExpressionType.COMBINED, "[all [of]] [the] loaded chunks (of|in) %world%");
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprLoadedChunks.java:43]: _Optional_
Can this support multiple worlds?

**Review by AyhamAl-Ali — APPROVED** [review]
Nice PR ⚡

File: src/main/java/ch/njol/skript/expressions/ExprLoadedChunks.java (hunk 2 above)
Comment (DelayedGaming) [reply src/main/java/ch/njol/skript/expressions/ExprLoadedChunks.java:43]: At first I thought that the usage of acquiring chunks from multiple world is low. I guess it is fine to support it so users don't have to loop, unlike #5401.

**Review by AyhamAl-Ali — APPROVED** [review]

**Comment by TheLimeGlass** [comment]
Should probably be included within https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/expressions/ExprChunk.java with separate syntax pattern `(all of the|all [the]) [loaded] chunks of %worlds%` and `[the] chunk[s] (of|%-directions%) %locations%`

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]
Should probably be included within https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/expressions/ExprChunk.java with separate syntax pattern `(all of the|all [the]) [loaded] chunks of %worlds%` and `[the] chunk[s] (of|%-directions%) %locations%`

**Comment by DelayedGaming** [comment]
Aware that tests are failing.

**Comment by DelayedGaming** [comment]
Fixed cause of test failing and PR ready for review!

**Review by AyhamAl-Ali — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprChunk.java [hunk 3]
```diff
@@ -18,83 +18,117 @@
...
+})
+@Since("2.0, INSERT VERSION (loaded chunks)")
+public class ExprChunk extends SimpleExpression<Chunk> {
 	
 	static {
-		Skript.registerExpression(ExprChunk.class, Chunk.class, ExpressionType.PROPERTY, "[the] chunk[s] (of|%-directions%) %locations%", "%locations%'[s] chunk[s]");
+		Skript.registerExpression(ExprChunk.class, Chunk.class, ExpressionType.COMBINED,
+			"[the] chunk[s] (of|%-directions%) %locations%",
+			"%locations%'[s] chunk[s]",
+			"[all [of]] [the] loaded chunks (of|in) %worlds%"
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprChunk.java:56]:
```suggestion
				"[all [of]] [the] chunk[s] (of|%-directions%) %locations%",
				"%locations%'[s] chunk[s]",
				"[all [of]] [the] loaded chunks (of|in) %worlds%"
```

File: src/main/java/ch/njol/skript/expressions/ExprChunk.java [hunk 4]
```diff
@@ -18,83 +18,117 @@
...
-	
+
 	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		return "the chunk at " + locations.toString(e, debug);
+	protected @Nullable Chunk[] get(Event event) {
+		if (pattern != 2) {
+			return locations.stream(event)
+				.map(Location::getChunk)
+				.toArray(Chunk[]::new);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprChunk.java:86]:
```suggestion
					.map(Location::getChunk)
					.toArray(Chunk[]::new);
```

File: src/main/java/ch/njol/skript/expressions/ExprChunk.java [hunk 5]
```diff
@@ -18,83 +18,117 @@
...
-	@Override
-	public Class<? extends Chunk> getReturnType() {
-		return Chunk.class;
-	}
-	
+
 	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		return "the chunk at " + locations.toString(e, debug);
+	protected @Nullable Chunk[] get(Event event) {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprChunk.java:82]:
```suggestion
	@Nullable
	protected Chunk[] get(Event event) {
```

File: src/main/java/ch/njol/skript/expressions/ExprChunk.java [hunk 6]
```diff
@@ -18,83 +18,117 @@
...
+				.toArray(Chunk[]::new);
+		}
+		World[] worlds = this.worlds.getArray(event);
+		if (worlds.length == 0)
+			return new Chunk[0];
+		List<Chunk> chunks = new ArrayList<>();
+		for (World world : worlds) {
+			chunks.addAll(Arrays.asList(world.getLoadedChunks()));
+		}
+		return chunks.toArray(new Chunk[0]);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprChunk.java:95]:
```suggestion
		return worlds.stream(event)
				.flatMap(world -> Arrays.stream(world.getLoadedChunks()))
				.toArray(Chunk[]::new);
```

File: src/main/java/ch/njol/skript/expressions/ExprChunk.java [hunk 7]
```diff
@@ -18,83 +18,117 @@
...
-	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) {
+	@SuppressWarnings("deprecation")
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
 		assert mode == ChangeMode.RESET;
-		
-		final Chunk[] cs = getArray(e);
-		for (final Chunk c : cs)
-			c.getWorld().regenerateChunk(c.getX(), c.getZ());
+		Chunk[] chunks = get(event);
+		for (Chunk chunk : chunks)
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprChunk.java:111]:
```suggestion
		for (Chunk chunk : get(event))
```

File: src/main/java/ch/njol/skript/expressions/ExprChunk.java [hunk 8]
```diff
@@ -18,83 +18,117 @@
...
+
+	@Override
+	public Class<? extends Chunk> getReturnType() {
+		return Chunk.class;
 	}
 	
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		if (pattern == 2)
+			return "the loaded chunks of " + worlds.toString(event, debug);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprChunk.java:130]:
```suggestion
			return "loaded chunks of " + worlds.toString(event, debug);
```

File: src/main/java/ch/njol/skript/expressions/ExprChunk.java [hunk 9]
```diff
@@ -18,83 +18,117 @@
...
+		return "the chunk at " + locations.toString(event, debug);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprChunk.java:131]:
```suggestion
		return "chunk at " + locations.toString(event, debug);
```
Skript can add `the` in some context already for printing.

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprChunk.java [hunk 10]
```diff
@@ -18,83 +18,109 @@
...
+@Examples({
+	"add the chunk at the player to {protected chunks::*}",
+	"set {_chunks::*} to the loaded chunks of the player's world"
+})
+@Since("2.0, INSERT VERSION (loaded chunks)")
+public class ExprChunk extends SimpleExpression<Chunk> {
 	
 	static {
-		Skript.registerExpression(ExprChunk.class, Chunk.class, ExpressionType.PROPERTY, "[the] chunk[s] (of|%-directions%) %locations%", "%locations%'[s] chunk[s]");
+			Skript.registerExpression(ExprChunk.class, Chunk.class, ExpressionType.COMBINED,
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprChunk.java:51]:
```suggestion
		Skript.registerExpression(ExprChunk.class, Chunk.class, ExpressionType.COMBINED,
```

**Review by TheLimeGlass — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprChunk.java [hunk 11]
```diff
@@ -18,83 +18,109 @@
...
+				"[all [of]] [the] chunk[s] (of|%-directions%) %locations%",
+				"%locations%'[s] chunk[s]",
+				"[all [of]] [the] loaded chunks (of|in) %worlds%"
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/expressions/ExprChunk.java:54]: I don't think this pattern looks quite right.
```
all loaded chunks (ok)
all the loaded chunks (ok)
all of the loaded chunks (ok)
all of loaded chunks (???)
the loaded chunks (ok)
```
You could try something like `(all [[of] the]|the|)` instead to eliminate the bad `all of chunks` pattern.

**Review by Moderocky — CHANGES_REQUESTED** [review]

**Review by Moderocky — APPROVED** [review]

**Comment by Moderocky** [comment]
This could have a test too, since spawn chunks ought to be loaded in the test environment.

**Comment by TheLimeGlass** [comment]
> This could have a test too, since spawn chunks ought to be loaded in the test environment.

too many variables

**Pull request merged**